notebooks/noj_book/echarts.clj

```diff
@@ -189,23 +227,164 @@ data-for-waterfall
               :itemStyle {:normal {:barBorderColor "rgba(0,0,0,0)"
                                    :color "rgba(0,0,0,0)"}
                           :emphasis {:barBorderColor "rgba(0,0,0,0)"
                                      :color "rgba(0,0,0,0)"}}
               :data suspended-cumulative}
              {:name "positive"
-              :type "bar"
+              :type :bar
               :stack "all"
               :data positive}
              {:name "negative"
-              :type "bar"
+              :type :bar
               :stack "all"
               :data negative
               :itemStyle {:color "#f33"}}]}))
+
+;; ### Line
+;; #### Dataset Preparation
+;; Before we plot any line charts, it's helpful to prepare the dataset first.
+
+(def stocks
+  (tc/dataset
+   "https://raw.githubusercontent.com/techascent/tech.ml.dataset/master/test/data/stocks.csv"
+   {:key-fn keyword}))
+
+;; This dataset originally contains three columns:
+(tc/head stocks)
+
+;; To make it better serve this tutorial, let's widen it:
+(def reshaped-stocks
+  (-> stocks
+      (tc/pivot->wider [:symbol] [:price] {:drop-missing? false})
+      (tc/rename-columns keyword)))
+
+reshaped-stocks
+
+;; As you can see, now it has a date column and a few other columns for every company's stock price.
+
+;; #### Basic Line Chart
+
+;; First let's try it out with some simple data manually:
+
+(kind/echarts
+ {:tooltip {}
+  :xAxis {:type :category
+          :data ["A" "B" "C"]}
+  :yAxis {:type :value}
+  :series [{:type :line
+            :data [120 200 150]}]})
+
+;; Now, we can try the prepared dataset to see how it goes:
+
+(kind/echarts
+ {:tooltip {}
+  :xAxis {:type :category
+          :data (tc/column reshaped-stocks :date)}
+  :yAxis {:type :value}
+  :series [{:type :line
+            :data (tc/column reshaped-stocks :MSFT)}]})
+
+;; #### Stacked Line Chart
+
+;; Now let's stack a few more lines onto the same chart by adding more data `series`:
+
+(kind/echarts
+ {:xAxis {:type :category
+          :data (tc/column reshaped-stocks :date)}
+  :yAxis {:type :value}
+  :series [{:type :line
+            :data (tc/column reshaped-stocks :MSFT)}
+           {:type :line
+            :data (tc/column reshaped-stocks :AMZN)}]})
+
+;; So far so good. But it's confusing without legend for these lines, so let's add it.
+
+(kind/echarts
+ {:xAxis {:type :category
+          :data (tc/column reshaped-stocks :date)}
+  :yAxis {:type :value}
+  :legend {:data [:MSFT :AMZN]}
+  :series [{:type :line
+            :data (tc/column reshaped-stocks :MSFT)
+            :name :MSFT}
+           {:type :line
+            :data (tc/column reshaped-stocks :AMZN)
+            :name :AMZN}]})
+
+;; Please note that you can toggle each line on and off by clicking its legend, which is a really nice feature.
+
+;; It's a bit tedious to have column names here and there,
+;; so we can define a helper function here to make life easier (maybe someday we can expand `kind` with this):
```

Comment:
(It is great that you are starting to create those helper functions. Probably, it will not be in Kindly but in a dedicated data visualization library.)